### ISSUE TITLE
glTF support is missing features

### ISSUE THREAD

**Issue by romainguy** [google/filament#528]
Our glTF viewer doesn't work entirely just yet. Here's what I found so far:
- Tangents/bitangents are not computed correctly if not present in the model
- We always use the default sampler instead of using the sampler described in the glTF file
- Vertex colors are ignored
- Transforms (rotation in particular) are not always handled properly

I found these issues while trying to load various Sketchfab assets.

**Comment by romainguy** [comment]
Rotation issue: https://sketchfab.com/models/a4ca0e3864e143af9bf1dffc7fc8029b (you have to zoom into the black sphere to seen the scene). This model also seems to use vertex colors? Our render is definitely off.

Non-default samplers: https://sketchfab.com/models/94b24a60dc1b48248de50bf087c0f042 The tramway and some of the walls are not textured properly, most likely because we don't use the right sampler

None of the textures are loaded for this model: https://sketchfab.com/models/282f497334f64a589edee4e63ad7e428

**Comment by romainguy** [comment]
Another bug: the water in https://sketchfab.com/models/e051a32d21d6453c8fad408082492267 is not transparent

**Comment by prideout** [comment]
I'm working on the first bullet in the original problem description (**Tangents/bitangents are not computed correctly**), which is exemplified by `glTF-Sample-Models/2.0/NormalTangentTest`.

**Comment by magicwhale** [comment]
> Rotation issue: https://sketchfab.com/models/a4ca0e3864e143af9bf1dffc7fc8029b (you have to zoom into the black sphere to seen the scene). This model also seems to use vertex colors? Our render is definitely off.
> 
> Non-default samplers: https://sketchfab.com/models/94b24a60dc1b48248de50bf087c0f042 The tramway and some of the walls are not textured properly, most likely because we don't use the right sampler
> 
> None of the textures are loaded for this model: https://sketchfab.com/models/282f497334f64a589edee4e63ad7e428

The first model seems to just have some rotation issues with the asset itself, and some of the colors aren't showing up because vertex colors aren't implemented yet.
For the second and third model, the issue seems to just be with the use of multiple UV coordinates.





> Another bug: the water in https://sketchfab.com/models/e051a32d21d6453c8fad408082492267 is not transparent. The alpha mask is also not working on the bushes in that scene.

I realized that I didn't implement alpha masking and blending for unlit materials, which this scene uses.

I'm working on fixing the issues now.

**Comment by romainguy** [comment]
Crash opening this file: https://sketchfab.com/models/f8d3c7506f3545108dcf7f6c170e19e1

**Comment by magicwhale** [comment]
It seems like that file and the files mentioned in #503  have some kind of error within the gltf file, and this is causing the viewer to crash within Assimp when it is loading the gltf. 

I saw that Don McCurdy's gltf viewer is still able to show the model even though it says there are errors within the file, so we could find a way to load the files and ignore the malformed lines.

**Comment by donmccurdy** [comment]
> It seems like that file and the files mentioned in #503 have some kind of error within the gltf file, and this is causing the viewer to crash within Assimp when it is loading the gltf.

If the errors seem unreasonable, I'm glad to send feedback to Sketchfab. Some viewers (like Facebook 3D Posts) enforce a "no validation errors" policy, and if these are issues you'd rather not work around, requiring stricter conformance from exporters is acceptable. The tolerance in my viewer is more accidental than planned. :)

**Comment by romainguy** [comment]
Well crashing is never a good option :) But if the crash is inside Assimp our best bet would be to fix it there and upstream the change. We could also investigate TinyGLTF

**Comment by romainguy** [comment]
Multi UV sets support is now in. @magicwhale were you able to figure out where the other issues come from?

**Comment by cdata** [comment]
@romainguy we recently proposed testing rendering parity between [`<model-viewer>`](https://github.com/GoogleWebComponents/model-viewer) (based on Three.js / GLTF Loader) and Filament for a number of the [Khronos sample glTFs](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0) (relevant [PR](https://github.com/GoogleWebComponents/model-viewer/pull/375)). We found samples that reproduce a number of the issues you referenced here. I'm leaving some of our test results below in case they might help inform your work to fix the related issues:

> Tangents/bitangents are not computed correctly if not present in the model

Related sample model: [Normal Tangent Mirror Test](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/NormalTangentMirrorTest)

Filament result|Expected result
---------------|---------------
![image](https://user-images.githubusercontent.com/240083/53101492-b7a5ef00-34de-11e9-8576-475952487008.png)|<img width="1536" src="https://user-images.githubusercontent.com/240083/53101603-f89e0380-34de-11e9-9a86-11fe6f291407.png">

> Vertex colors are ignored

Related sample model: [Vertex Color Test](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/VertexColorTest)

Filament result|Expected result
---------------|---------------
![image](https://user-images.githubusercontent.com/240083/53101798-52063280-34df-11e9-91cd-89d149a88e8e.png)|<img width="1536" src="https://user-images.githubusercontent.com/240083/53101759-3d299f00-34df-11e9-9dfc-11d9224addda.png">

> Transforms (rotation in particular) are not always handled properly

Related sample model: [Texture Transform Test](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TextureTransformTest)

When I saw the results of this test it rang a bell. I'm not sure if you meant texture rotations, but perhaps this is relevant.

Filament result|Expected result
---------------|---------------
![image](https://user-images.githubusercontent.com/240083/53102102-e5d7fe80-34df-11e9-875d-615f02b670e4.png)|<img width="1536" src="https://user-images.githubusercontent.com/240083/53102121-ee303980-34df-11e9-8b25-8c0b3eb197eb.png">

**Comment by romainguy** [comment]
Thank you. Texture transforms are just not supported at all at the moment. We're moving to a different gltf loader to fix all these issues.

**Comment by romainguy** [comment]
I'm surprised by the tangents though, @prideout wasn't this fixed?

**Comment by prideout** [comment]
Just now reproduced the issue with Normal Tangent Mirror, I'm surprised too.  Will investigate.

**Comment by prideout** [comment]
tl;dr:

For now I think I should simply modify our copy of MeshAssimp to negate the bitangent, thus moving [our current workaround](https://github.com/google/filament/blob/26ce90483a1f135cdd1e189a362a709416680578/samples/app/MeshAssimp.cpp#L845-L850) downward so that it doesn't affect the tangents-are-available case.  This will be fixed in a more proper way in the new glTF loader that I'm working on.

details:

We render `NormalTangentTest` correctly (where tangents are computed via MeshAssimp) but we do not render `NormalTangentMirrorTest` correctly (where tangents are supplied in the model).

When MeshAssimp generates tangents and bitangents, it basically makes tangents follow texcoord U, and bitangents follow texcoord V.  The resulting tangent vectors are fairly consistent with a couple popular viewers that I tried, but the bitangent vectors are not.  Other viewers use different approaches:

- ThreeJS (and therefore Don McCurdy's viewer) does [this](https://github.com/mrdoob/three.js/blob/f89180d3df9021121bc1d670f1223ff2166eddc2/src/renderers/shaders/ShaderLib/normal_vert.glsl.js#L46)
- The Khronos ref viewer does [this](https://github.com/KhronosGroup/glTF-WebGL-PBR/blob/88eda8c5358efe03128b72b6c5f5f6e5b6d023e1/shaders/pbr-frag.glsl#L112-L126)
- The spec recommends mikktspace (which Blender uses I think) which does [this](https://github.com/tcoppex/ext-mikktspace/blob/68ae624c995a0018e6e04f2a397685324571284e/mikktspace.c#L977-L978)

**Comment by donmccurdy** [comment]
> ThreeJS (and therefore Don McCurdy's viewer) does [this](https://github.com/mrdoob/three.js/blob/f89180d3df9021121bc1d670f1223ff2166eddc2/src/renderers/shaders/ShaderLib/normal_vert.glsl.js#L46)

Note that this change is [very recent](https://github.com/mrdoob/three.js/pull/15749), and won't be visible in my viewer until threejs r102 is released. But I think BabylonJS (and [this viewer](http://sandbox.babylonjs.com/)) do the same.

**Comment by romainguy** [comment]
Closing this for now. There are still a couple of unsupported features like blend shapes.
